streams/src/test/java/org/apache/kafka/streams/smoketest/BrokerCompatibilityTest.java

```diff
@@ -0,0 +1,100 @@
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.kstream.KStreamBuilder;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.Properties;
+
+public class BrokerCompatibilityTest {
+
+    public static final String SOURCE_TOPIC = "brokerCompatibilitySourceTopic";
+    public static final String SINK_TOPIC = "brokerCompatibilitySinkTopic";
+
+    public static void main(String[] args) throws Exception {
+        System.out.println("StreamsTest instance started");
+
+        final String kafka = args.length > 0 ? args[0] : "localhost:9092";
+        final String stateDirStr = args.length > 1 ? args[1] : "/tmp/kafka-streams-system-test";
+
+        final File stateDir = new File(stateDirStr);
+        stateDir.mkdir();
+
+        final Properties streamsProperties = new Properties();
+        streamsProperties.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        streamsProperties.put(StreamsConfig.APPLICATION_ID_CONFIG, "kafka-streams-system-test-broker-compatibility");
+        streamsProperties.put(StreamsConfig.STATE_DIR_CONFIG, stateDir.toString());
+        streamsProperties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        streamsProperties.put(StreamsConfig.KEY_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+        streamsProperties.put(StreamsConfig.VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+        streamsProperties.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100);
+
+
+        final KStreamBuilder builder = new KStreamBuilder();
+        builder.stream(SOURCE_TOPIC).to(SINK_TOPIC);
+
+        final KafkaStreams streams = new KafkaStreams(builder, streamsProperties);
+        streams.start();
+
+
+        final Properties producerProperties = new Properties();
+        producerProperties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        producerProperties.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        producerProperties.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+
+        final KafkaProducer<String, String> producer = new KafkaProducer<>(producerProperties);
+        producer.send(new ProducerRecord<>(SOURCE_TOPIC, "key", "value"));
+
+
+        final Properties consumerProperties = new Properties();
+        consumerProperties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        consumerProperties.put(ConsumerConfig.GROUP_ID_CONFIG, "broker-compatibility-consumer");
+        consumerProperties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        consumerProperties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        consumerProperties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+
+        final KafkaConsumer<String, String> consumer = new KafkaConsumer<>(consumerProperties);
+        consumer.subscribe(Collections.singletonList(SINK_TOPIC));
+
+        outer: while (true) {
```

Comment:
Can't say i like using labels and breaking to them. Feels like a dirty hack to me. I'd prefer that you either loop on a condition or extract the loop to a method and return from it when it is done.